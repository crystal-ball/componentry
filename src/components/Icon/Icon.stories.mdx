import { ArgsTable, Meta, Story } from '@storybook/addon-docs'

import { Icon } from './Icon'

<Meta title='Components/Icon' component={Icon} />

# Icon

## Variants

Componentry ships with a single `font` variant which creates font icons:

<Story name='Icon'>
  <Icon id='coffee' />
</Story>

## Icon alignment

Icons are center aligned in flex and text contexts by default for convenience:

In flex containers icons are centered with `self-align: center`:

<h1 className='flex gap-0.5'>
  <Icon id='code' />
  Flex alignment
</h1>

<h2 className='flex gap-0.5'>
  <Icon id='code' />
  Flex alignment
</h2>

<h3 className='flex gap-0.5'>
  <Icon id='code' />
  Flex alignment
</h3>

In text elements icons are centered with `vertical-align: -0.15em`:

<h1>
  <Icon id='code' /> Flex alignment
</h1>

<h2>
  <Icon id='code' /> Flex alignment
</h2>

<h3>
  <Icon id='code' /> Flex alignment
</h3>

Checkmate... Hey, you know how I'm, like, always trying to save the planet? Here's my
chance <Icon id="code" matchTextBaseline />.They're using our own satellites against us.
And the clock is ticking. Must go faster... go, go, go, go, go! I was part of something
special. Remind me to thank John for a lovely weekend.

## Props

<ArgsTable of={Icon} />
