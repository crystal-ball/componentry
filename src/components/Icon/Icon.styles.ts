import { StylesDefinition } from '../../utils/types'

//                                         <Icon /> styles
// -------------------------------------------------------

export const iconStyles: IconStyles = {
  // BASE
  '.C9Y-Icon-base': {
    display: 'inline-block',
    userSelect: 'none',
  },

  // VARIANTS
  '.C9Y-Icon-font': {
    // 1em width+height makes icons font-sized by default ðŸ”®
    height: '1em',
    width: '1em',
    // Helpful default to help prevent icon from getting squished
    flexShrink: 0,
    // Alignment: In flex layouts default to centered
    alignSelf: 'center',
    // Alignment: Outside flex layouts center to text baseline with negative vertical align
    verticalAlign: '-.15em',
    // Default to use the parent color as the icon color
    fill: 'currentColor',
  },

  // SIZES
  // ...coming soon
}

export interface IconStyles {
  /** Base class applied to all variants for shared structural styles */
  '.C9Y-Icon-base': StylesDefinition
  /** Variant class applied when `variant="font"` */
  '.C9Y-Icon-font': StylesDefinition
}
