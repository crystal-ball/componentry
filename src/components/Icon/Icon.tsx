import React, { forwardRef } from 'react'
import { element } from '../../utils/element-creator'
import { MergeTypes, Resolve } from '../../utils/types'
import { UtilityProps } from '../../utils/utility-classes'
import { useThemeProps } from '../Provider/Provider'

// --------------------------------------------------------
// ICON ELEMENTS MAP

/** Mapping of icon IDs to components rendered by Icon */
export type IconElementsMap = { [ID: string]: React.ComponentType<unknown> }

let iconElementsMap: IconElementsMap = {}

/**
 * Configuration method for defining the elements to render for each Icon ID.
 * @remarks
 * Configuring an icon elements map isn't necessary if you've setup an SVG
 * symbol sprite.
 * @example
 * ```ts
 * import Info from './info.svg'
 * import Coffee from './coffee.svg'
 *
 * configureIconElementsMap({
 *   info: Info,
 *   coffee: Coffee
 * })
 * ```
 */
export function configureIconElementsMap(elementsMap: IconElementsMap): void {
  iconElementsMap = elementsMap
}

// --------------------------------------------------------
// ICON COMPONENT

/** Module augmentation interface for overriding component props' types */
export interface IconPropsOverrides {}

export interface IconPropsDefaults {
  /** External path to symbol sprite  */
  externalURI?: string
  /** ID for the `iconElementsMap` or href attribute for symbol sprites */
  id: string
  /** Display variant */
  variant?: 'font'
}

export type IconProps<Elem extends React.ElementType = 'svg'> = Resolve<
  MergeTypes<IconPropsDefaults, IconPropsOverrides>
> &
  UtilityProps &
  React.ComponentPropsWithRef<Elem> & { as?: Elem }

/**
 * Icon provides consistently themed iconography elements.
 * @example
 * ```tsx
 * <Icon id="coffee" />
 * ```
 * @see [üìù Icon](https://componentry.design/docs/components/icon)
 */
export interface Icon {
  <Elem extends React.ElementType = 'svg'>(
    props: IconProps<Elem>,
  ): React.ReactElement | null
  displayName?: string
}

export const Icon: Icon = forwardRef<HTMLElement, IconProps>((props, ref) => {
  const {
    externalURI = '',
    id,
    variant = 'font',
    ...rest
  } = { ...useThemeProps('Icon'), ...props }

  const hasMappedElement = id in iconElementsMap

  return element({
    ref,
    as: hasMappedElement ? iconElementsMap[id] : 'svg',
    children: hasMappedElement ? undefined : <use href={`${externalURI}#${id}`} />,
    componentCx: `C9Y-Icon-base C9Y-Icon-${variant} icon-${id}`,
    role: 'img',
    'aria-label': id,
    ...rest,
  })
})
Icon.displayName = 'Icon'
