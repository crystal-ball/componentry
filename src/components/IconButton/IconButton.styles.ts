import { getMergedConfig } from '../../plugin-postcss/configs'
import { StylesDefinition } from '../../utils/types'

const { theme } = getMergedConfig()

//                                        <Button /> styles
// --------------------------------------------------------

export const iconButtonStyles: IconButtonStyles = {
  // BASE
  '.C9Y-IconButton-base': {
    alignItems: 'center',
    display: 'inline-flex',
    justifyContent: 'center',
    gap: '0.5rem',
    fontSize: '0.875rem',
    lineHeight: 1, // ensures text is center aligned within flex layout

    border: 'none', // reset borders for easier opt-in styling
    userSelect: 'none', // Prevent text selection on click of buttons
    whiteSpace: 'nowrap', // By default button content shouldn't wrap
  },

  '.C9Y-IconButton-DisabledWrapper': {
    cursor: 'not-allowed',
  },

  // VARIANTS
  '.C9Y-IconButton-filled': {
    height: '2rem',
    width: '2rem',
    backgroundColor: theme.colors.primary[500],
    borderRadius: theme.borderRadius.DEFAULT,
    color: theme.colors.inverse,

    transition: 'background-color 0.15s linear',

    '&:hover, &.C9Y-hover': {
      backgroundColor: theme.colors.primary[700],
    },
    '&:active, &.C9Y-active': {
      backgroundColor: theme.colors.primary[900],
    },
    '&:disabled, &.C9Y-disabled': {
      backgroundColor: theme.colors.primary[300],
    },
  },

  '.C9Y-IconButton-outlined': {
    height: '2rem',
    width: '2rem',
    backgroundColor: 'transparent',
    border: `1px solid ${theme.colors.primary[500]}`,
    borderRadius: theme.borderRadius.DEFAULT,
    color: theme.colors.primary[500],

    transition: 'color, border-color 0.15s linear',

    '&:hover, &.C9Y-hover': {
      borderColor: theme.colors.primary[700],
      color: theme.colors.primary[700],
    },
    '&:active, &.C9Y-active': {
      borderColor: theme.colors.primary[900],
      color: theme.colors.primary[900],
    },
    '&:disabled, &.C9Y-disabled': {
      borderColor: theme.colors.primary[300],
      color: theme.colors.primary[300],
    },
  },

  // SIZES
  '.C9Y-IconButton-smallSize': {
    height: '1.5rem',
    width: '1.5rem',
    fontSize: '0.75rem',
  },
  '.C9Y-IconButton-largeSize': {
    height: '2.5rem',
    width: '2.5rem',
    fontSize: '1rem',
  },
}

export interface IconButtonStyles {
  /** Base class applied to all variants for shared structural styles */
  '.C9Y-IconButton-base': StylesDefinition
  /** Class applied to disabled button wrapper element */
  '.C9Y-IconButton-DisabledWrapper': StylesDefinition

  /** Variant class applied when `variant="transparent"` */
  '.C9Y-IconButton-filled': {
    '&:hover, &.C9Y-hover': StylesDefinition
    '&:active, &.C9Y-active': StylesDefinition
    '&:disabled, &.C9Y-disabled': StylesDefinition
  } & StylesDefinition
  /** Variant class applied when `variant="outlined"` */
  '.C9Y-IconButton-outlined': {
    '&:hover, &.C9Y-hover': StylesDefinition
    '&:active, &.C9Y-active': StylesDefinition
    '&:disabled, &.C9Y-disabled': StylesDefinition
  } & StylesDefinition
  /** Sizing class applied when `size="small"` */
  '.C9Y-IconButton-smallSize': StylesDefinition
  /** Sizing class applied when `size="large"` */
  '.C9Y-IconButton-largeSize': StylesDefinition
}
