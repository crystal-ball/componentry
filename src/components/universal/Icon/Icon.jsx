// @flow
import React, { Component } from 'react'
import classNames from 'classnames'

type Props = {
  className: ?string,
  font: ?boolean,
  id: string,
  title: ?string,
  presentation: ?boolean,
}

/**
 * Use the icon component with an svg imported from `/media/icons` to create an icon
 * font.
 */
export default class Icon extends Component<Props> {
  static defaultProps = {
    className: '',
    font: true,
    presentation: false,
    title: null,
  }

  constructor(props: Props) {
    super(props)
    // $FlowIgnore: Flow doesn't know about webpack magic comments for dynamic imports
    import(/* webpackMode: "eager", webpackChunkName: "icons" */ `media/icons/${
      props.id
    }.svg`)
  }

  render() {
    const { title, presentation, id, className, font } = this.props

    return (
      <svg
        aria-label={title}
        role={presentation ? 'presentation' : 'img'}
        className={classNames('icon', id, className, { font })}
      >
        <use xlinkHref={`#${id}`} href={`#${id}`} />
      </svg>
    )
  }
}
