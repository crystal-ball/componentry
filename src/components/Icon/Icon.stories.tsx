import type { Meta, StoryObj } from '@storybook/react'

import { Icon } from './Icon'

const meta: Meta<typeof Icon> = {
  component: Icon,
}

export default meta
type Story = StoryObj<typeof Icon>

export const Primary: Story = {
  args: {
    id: 'coffee',
  },
}
