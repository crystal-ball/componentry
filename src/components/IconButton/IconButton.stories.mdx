import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs'
import { Icon } from '../Icon/Icon'
import { IconButton } from './IconButton'

<Meta title='Components/IconButton' component={IconButton} />

export const Template = (args) => <IconButton {...args} />

# IconButton

## Variants

Pass a `filled` or `outlined` variant to set a button style.

<Story
  name='IconButton'
  decorators={[
    (Story) => (
      <div className='grid grid-cols-4 gap-x-4 gap-y-2 items-center justify-items-center w-1/2'>
        <div>base</div>
        <div>hover</div>
        <div>active</div>
        <div>disabled</div>
        <Story />
      </div>
    ),
  ]}
>
  <IconButton variant='filled' icon={<Icon id='code' />} />
  <IconButton variant='filled' icon={<Icon id='code' />} className='C9Y-hover' />
  <IconButton variant='filled' icon={<Icon id='code' />} className='C9Y-active' />
  <IconButton variant='filled' icon={<Icon id='code' />} disabled />
  <IconButton variant='outlined' icon={<Icon id='code' />} />
  <IconButton variant='outlined' icon={<Icon id='code' />} className='C9Y-hover' />
  <IconButton variant='outlined' icon={<Icon id='code' />} className='C9Y-active' />
  <IconButton variant='outlined' icon={<Icon id='code' />} disabled />
</Story>

## Sizes

Pass a `size` to set a button size:

<div className='flex flex-col gap-4 justify-center'>
  <div className='flex flex-row items-center gap-4'>
    <IconButton variant='filled' icon={<Icon id='code' />} size='small' />
    <IconButton variant='filled' icon={<Icon id='code' />} />
    <IconButton variant='filled' icon={<Icon id='code' />} size='large' />
  </div>
  <div className='flex flex-row items-center gap-4'>
    <IconButton variant='outlined' icon={<Icon id='code' />} size='small' />
    <IconButton variant='outlined' icon={<Icon id='code' />} />
    <IconButton variant='outlined' icon={<Icon id='code' />} size='large' />
  </div>
</div>

## Disabled

- Controlled by `disabled` prop
- Adds `disabled` attribute to element and `.C9Y-Disabled` class
- Adds `pointer-events: none` to prevent user actions

<div className='flex items-center gap-4'>
  <IconButton icon={<Icon id='code' />} disabled />
</div>

## IconButtons with hrefs

IconButtons with an `href` will be rendered as anchors:

<div className='flex flex-col items-center gap-4'>
  <IconButton icon={<Icon id='code' />} href='#test' />
  <IconButton disabled icon={<Icon id='code' />} href='#test' />
</div>

<ArgsTable of={IconButton} />
