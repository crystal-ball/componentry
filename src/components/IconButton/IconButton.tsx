import React, { forwardRef } from 'react'
import { element } from '../../utils/element-creator'
import { MergeTypes, Resolve } from '../../utils/types'
import { UtilityProps } from '../../utils/utility-classes'
import { Icon } from '../Icon/Icon'
import { useThemeProps } from '../Provider/Provider'

// Module augmentation interface for overriding component props' types
export interface IconButtonPropsOverrides {}

export interface IconButtonPropsDefaults {
  /** Display variant color */
  color?: 'primary'
  /** Disables the element, preventing mouse and keyboard events */
  disabled?: boolean
  /** Button content */
  icon: string | JSX.Element
  /** Toggles full width element layout */
  fullWidth?: boolean
  /** HTML element href */
  href?: string
  /** Sets the display size */
  size?: 'small' | 'large'
  /** Indicates whether buttons in a disabled state should be wrapped with a span */
  wrapWhenDisabled?: boolean
  /** Display variant */
  variant?: 'transparent' | 'outlined'
}

export type IconButtonProps = Resolve<
  MergeTypes<IconButtonPropsDefaults, IconButtonPropsOverrides>
> &
  UtilityProps &
  React.ComponentPropsWithRef<'button'>

// ‚ú® Nice display type for IntelliSense
export interface IconButton {
  (props: IconButtonProps): React.ReactElement | null
  displayName?: string
}

/**
 * #### [üìù IconButton](https://componentry.design/docs/components/iconbutton)
 *
 * IconButton provides an accessible icon action element
 * @example
 * ```tsx
 * <IconButton id="code" onClick={() => buildSomethingRadical()} />
 * ```
 */
export const IconButton: IconButton = forwardRef((props, ref) => {
  const {
    variant = 'filled',
    color,
    disabled,
    icon,
    size,
    wrapWhenDisabled = true,
    ...merged
  } = {
    ...useThemeProps('IconButton'),
    ...props,
  }

  const contents = element({
    ref,
    disabled,
    // If an href is passed, this instance should render an anchor tag
    as: merged.href ? 'a' : 'button',
    type: merged.href ? undefined : 'button',
    componentCx: [
      `C9Y-IconButton-base C9Y-IconButton-${variant}`,
      {
        [`C9Y-IconButton-${color}Color`]: color,
        [`C9Y-IconButton-${size}Size`]: size,
      },
    ],
    children: typeof icon === 'string' ? <Icon id={icon} /> : icon,
    ...merged,
  })

  return disabled && wrapWhenDisabled ? (
    <span className='C9Y-IconButton-DisabledWrapper'>{contents}</span>
  ) : (
    contents
  )
})
IconButton.displayName = 'IconButton'
