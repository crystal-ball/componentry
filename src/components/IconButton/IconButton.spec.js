import { render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { Icon } from '../Icon/Icon'
import { IconButton } from './IconButton'

describe('<IconButton/>', () => {
  // Basic library element test suite
  elementTests(IconButton)

  it('When no props are passed, then defaults should be rendered', () => {
    render(<IconButton icon={<Icon id='code' />} />)

    // By default the button should have type button for a11y
    expect(screen.getByRole('button')).toHaveAttribute('type', 'button')
    // By default the variant primary
    expect(screen.getByRole('button')).toHaveClass(
      'C9Y-IconButton-base C9Y-IconButton-filled',
    )
  })

  it('When `type` is passed, then it overrides the default', () => {
    render(<IconButton icon={<Icon id='code' />} type='submit' />)

    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit')
  })

  it('When `variant` is passed, then it should be used as base className value', () => {
    render(<IconButton variant='outlined' icon={<Icon id='code' />} />)

    expect(screen.getByRole('button')).toHaveClass(
      'C9Y-IconButton-base C9Y-IconButton-outlined',
    )
  })

  it('When `color` is passed, then the color className should render', () => {
    render(<IconButton icon={<Icon id='code' />} color='primary' />)

    expect(screen.getByRole('button')).toHaveClass(
      'C9Y-IconButton-base  C9Y-IconButton-filled C9Y-IconButton-primaryColor',
    )
  })

  it('When `size` is passed, then the size className should render', () => {
    render(<IconButton icon={<Icon id='code' />} size='small' />)

    expect(screen.getByRole('button')).toHaveClass(
      'C9Y-IconButton-base C9Y-IconButton-smallSize',
    )
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<IconButton /> Snapshots', () => {
  it('renders defaults correctly', () => {
    render(<IconButton icon={<Icon id='code' />} />)

    expect(screen.getByRole('button')).toMatchSnapshot()
  })
})
