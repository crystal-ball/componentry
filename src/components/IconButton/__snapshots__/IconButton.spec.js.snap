// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> Snapshots renders defaults correctly 1`] = `
<button
  class="C9Y-IconButton-base C9Y-IconButton-filled"
  type="button"
>
  <svg
    aria-label="code"
    class="C9Y-Icon-base C9Y-Icon-font icon-code"
    role="img"
  >
    <use
      href="#code"
    />
  </svg>
</button>
`;
